qa/tasks/vstart_runner.py

```diff
@@ -434,8 +428,39 @@ def _do_run(self, args, check_status=True, wait=True, stdout=None,
             if os.path.exists(local_bin):
                 args = [local_bin] + args[1:]
 
+        # We have to use shell=True if any run.Raw was present, e.g. &&
+        if not shell:
+            shell = any([a for a in args if isinstance(a, Raw)])
+
         log.debug('> ' +
-                 ' '.join([str(a.value) if isinstance(a, Raw) else a for a in args]))
+                  ' '.join([str(a.value) if isinstance(a, Raw) else a for a in args]))
+
+        return args
+
+    def _perform_checks(self, args, omit_sudo, shell):
+        if isinstance(args, str):
+            return self._perform_checks_when_str(args, omit_sudo, shell)
+        elif isinstance(args, list):
+            return self._perform_checks_when_list(args, omit_sudo, shell)
+
+    # Wrapper to keep the interface exactly same as that of
+    # teuthology.remote.run.
+    def run(self, **kwargs):
+        """
+        Prepare arguments and get them executed.
+
+        We might have multiple commands in kwargs['args'] as list within list
+        or as string with newlines.
+        """
+        return self._do_run(**kwargs)
+
+    # XXX: omit_sudo is set to True since using sudo can change the ownership
+    # of files which becomes problematic for following executions of
+    # vstart_runner.py.
+    def _do_run(self, args, check_status=True, wait=True, stdout=None,
+                stderr=None, cwd=None, stdin=None, logger=None, label=None,
+                env=None, timeout=None, omit_sudo=True, shell=True):
+        args = self._perform_checks(args, omit_sudo, shell)
 
```

Comment:
```suggestion
prefix = """
adjust-ulimits() {
  "$@"
}
ceph-coverage() {
  "$@"
}
"""
cmd = prefix + quote(args)
```

Reply:
IMO this makes the log/output much less readable, compact and pretty -

```
}
ceph-coverage() {
  "$@"
}
./bin/ceph osd pool rm cephfs_data cephfs_data --yes-i-really-really-mean-it
2021-07-15T03:05:06.036 DEBUG:__main__:>
adjust-ulimits() {
  "$@"
}
ceph-coverage() {
  "$@"
}
./bin/ceph config rm mon mon_allow_pool_delete
2021-07-15T03:05:06.514 DEBUG:__main__:>
adjust-ulimits() {
  "$@"
}
ceph-coverage() {
  "$@"
}
```

I tried reducing number of lines it consumes but my opinion didn't change -

```
2021-07-15T03:06:00.323 DEBUG:__main__:> adjust-ulimits() { "$@"; }
ceph-coverage() { "$@"; }
./bin/ceph osd pool rm cephfs_metadata cephfs_metadata --yes-i-really-really-mean-it
2021-07-15T03:06:01.384 DEBUG:__main__:> adjust-ulimits() { "$@"; }
ceph-coverage() { "$@"; }
./bin/ceph osd pool rm cephfs_data cephfs_data --yes-i-really-really-mean-it
2021-07-15T03:06:02.450 DEBUG:__main__:> adjust-ulimits() { "$@"; }
ceph-coverage() { "$@"; }
./bin/ceph config rm mon mon_allow_pool_delete
2021-07-15T03:06:02.897 DEBUG:__main__:> adjust-ulimits() { "$@"; }
ceph-coverage() { "$@"; }
```

Deleting these commands is best option IMO. We can add warning for deletion of those commands at the very beginning of vstart_runner.py's output, in docstrings and add a warning in the documentation as well - https://docs.ceph.com/en/latest/dev/developer_guide/running-tests-locally/#running-test-using-vstart-runner-py.

Reply:
There's  no reason you can prepend those function hooks prior to invoking `/bin/sh`? I don't see why it needs to end up in the log.

Reply:
Deleting these commands is best option IMO. We can add warning for deletion of those commands at the very beginning of vstart_runner.py's output, in docstrings and add a warning in the documentation as well - https://docs.ceph.com/en/latest/dev/developer_guide/running-tests-locally/#running-test-using-vstart-runner-py.